lib/llvmopencl/linker.cpp

```diff
@@ -427,13 +436,153 @@ int copyKernelFromBitcode(const char* name, llvm::Module *parallel_bc,
-    parallel_bc->eraseNamedMetadata(DebugCU);
+  if (pocl_get_bool_option("POCL_LLVM_ALWAYS_INLINE", 0)) {
+    llvm::Module::iterator MI, ME;
+    for (MI = parallel_bc->begin(), ME = parallel_bc->end(); MI != ME; ++MI) {
+      Function *F = &*MI;
+      if (F->isDeclaration())
+          continue;
+      // inline all except the kernel
+      if (F->getName() != name) {
+          F->addFnAttr(Attribute::AlwaysInline);
+      }
+    }
 
+    llvm::legacy::PassManager Passes;
+    Passes.add(createAlwaysInlinerLegacyPass());
+    Passes.run(*parallel_bc);
+  }
   return 0;
 }
 
+bool moveProgramScopeVarsOutOfProgramBc(llvm::LLVMContext *Context,
+                                        llvm::Module *ProgramBC,
+                                        llvm::Module *OutputBC) {
+
+  ValueToValueMapTy VVM;
+  llvm::Module::global_iterator GI, GE;
+
+  // Copy all the globals from input to output module.
+  DB_PRINT("cloning the global variables:\n");
+  for (GI = ProgramBC->global_begin(), GE = ProgramBC->global_end(); GI != GE;
+       GI++) {
+    DB_PRINT(" %s\n", GI->getName().data());
+    GlobalVariable *GV = new GlobalVariable(
+        *OutputBC, GI->getValueType(), GI->isConstant(), GI->getLinkage(),
+        (Constant *)0, GI->getName(), (GlobalVariable *)0,
+        GI->getThreadLocalMode(), GI->getType()->getAddressSpace());
+    GV->copyAttributesFrom(&*GI);
+    // for program scope vars, change linkage to external
+    if (isProgramScopeVariable(*GV)) {
+      GV->setDSOLocal(false);
+      GV->setLinkage(GlobalValue::LinkageTypes::ExternalLinkage);
+    }
+    VVM[&*GI] = GV;
+  }
+
+  // add initializers to global vars, copy aliases & MD
+  std::string log;
+  shared_copy(OutputBC, ProgramBC, log, VVM);
+
+  // change program-scope variables in Input module to references
+  std::list<GlobalVariable *> GVarsToErase;
+  std::list<GlobalVariable *> ProgramGVars;
+  for (GI = ProgramBC->global_begin(), GE = ProgramBC->global_end(); GI != GE;
+       GI++) {
+    ProgramGVars.push_back(&*GI);
+  }
+
+  // we can't use iteration over ProgramBC->global_begin/end here, because we're
+  // adding new GVars into the Module during iteration
+  for (auto GI : ProgramGVars) {
+    DB_PRINT(" %s\n", GI->getName().data());
+    if (isProgramScopeVariable(*GI)) {
+      std::string OrigName = GI->getName().str();
+      GI->setName(Twine(OrigName, "__old"));
+      GlobalVariable *NewGV = new GlobalVariable(
+          *ProgramBC, GI->getValueType(), GI->isConstant(),
+          GlobalValue::LinkageTypes::ExternalLinkage,
+          (Constant *)nullptr, // initializer
+          OrigName,
+          (GlobalVariable *)nullptr, // insert-before
+          GI->getThreadLocalMode(), GI->getType()->getAddressSpace());
+
+      NewGV->copyAttributesFrom(GI);
+      NewGV->setDSOLocal(false);
+      GI->replaceAllUsesWith(NewGV);
+      GVarsToErase.push_back(GI);
+    }
+  }
+  for (auto GV : GVarsToErase) {
+    GV->eraseFromParent();
+  }
+
+#if 0
```

Comment:
Why this is disabled? Perhaps remove if it's not needed anymore.

Reply:
fixed